central/scannerdefinitions/handler/mappingUpdater.go

```diff
@@ -0,0 +1,154 @@
+)
+
+// newMappingUpdater creates a new updater for RH repository mapping data.
+func newMappingUpdater(file *file.File, client *http.Client, downloadURL string, interval time.Duration) *mappingUpdater {
+	if downloadURL == "" {
+		downloadURL = baseURL
+	}
+	return &mappingUpdater{
+		file:        file,
+		client:      client,
+		downloadURL: downloadURL,
+		interval:    interval,
+		stopSig:     concurrency.NewSignal(),
+	}
+}
+
+// Stop stops the updater.
+func (u *mappingUpdater) Stop() {
+	u.stopSig.Signal()
+}
+
+// Start starts the updater.
+// The updater is only started once.
+func (u *mappingUpdater) Start() {
+	u.once.Do(func() {
+		// Run the first update in a blocking-manner.
+		err := u.update()
+		if err != nil {
+			log.Errorf("Failed to start Scanner v4 repository mapping updater: %v", err)
+		}
+		go u.runForever()
+	})
+}
+
+func (u *mappingUpdater) OpenFile() (*os.File, time.Time, error) {
+	return u.file.Open()
+}
+
+func (u *mappingUpdater) runForever() {
+	timer := time.NewTimer(u.interval)
+	defer timer.Stop()
+
+	for {
+		select {
+		case <-timer.C:
+			err := u.update()
+			if err != nil {
+				log.Errorf("Failed to update Scanner v4 repository mapping: %v", err)
+			}
+			// Reset the timer with a new interval
+			timer.Reset(u.interval + nextInterval())
+		case <-u.stopSig.Done():
+			return
+		}
+	}
+}
+
+func (u *mappingUpdater) update() error {
+	if err := u.doUpdate(); err != nil {
+		log.Errorf("Failed to update Scanner v4 repository mapping from endpoint %q: %v", u.downloadURL, err)
+		return err
+	}
+	return nil
+}
+
+func (u *mappingUpdater) doUpdate() error {
+	err := u.downloadFromURL(u.downloadURL)
+	if err != nil {
+		return err
+	}
+
+	log.Info("Finished downloading repo mapping data for Scanner V4")
+	return nil
+}
+
+func (u *mappingUpdater) downloadFromURL(url string) error {
+	download := func() error {
+		resp, err := http.Get(url)
+		if err != nil {
+			return err
+		}
+
+		defer func() {
+			if err := resp.Body.Close(); err != nil {
+				log.Errorf("Error closing response body: %v", err)
+			}
+		}()
+
+		if resp.StatusCode != http.StatusOK {
+			return fmt.Errorf("received non-200 status code: %d", resp.StatusCode)
+		}
+
+		lastModified, err := time.Parse(time.RFC1123, resp.Header.Get(lastModifiedHeader))
+		if err != nil {
+			return errors.Errorf("unable to determine upstream definitions file's modified time: %v", err)
+		}
+		err = u.file.Write(resp.Body, lastModified)
+		if err != nil {
+			return err
+		}
+
+		return nil // Success case
+	}
+
+	// Notify function will log the error and the backoff delay duration
+	notify := func(err error, duration time.Duration) {
+		log.Errorf("Error: %v. Retrying in %v...\n", err, duration)
+	}
+
+	b := backoff.NewExponentialBackOff()
+	backoff.WithMaxRetries(b, 3) // Set max retry attempts to 3
+
+	return backoff.RetryNotify(download, b, notify)
+}
+
+func nextInterval() time.Duration {
+	addMinutes := []int{10, 20, 30, 40}
```

Comment:
I talked with @jvdm and he mentioned that in ACSCS there's memory spike for each 4 hours because the updaters work in that interval. Therefore he was suggesting we can avoid the updaters to work at the same time for less memory consumption.